src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSign1Message.cs

```diff
@@ -297,11 +305,56 @@ private async Task<bool> VerifyAsyncCore(AsymmetricAlgorithm key, Stream content
 
             HashAlgorithmName hashAlgorithm = GetHashAlgorithmFromCoseAlgorithmAndKeyType(nullableAlg.Value, keyType);
 
-            byte[] toBeSignedHash = new byte[GetHashSizeInBytes(hashAlgorithm)];
-            int bytesWritten = await CreateHashedToBeSignedAsync(toBeSignedHash, SigStructureCoxtextSign1, _protectedHeaderAsBstr, content, hashAlgorithm).ConfigureAwait(false);
-            Debug.Assert(bytesWritten == toBeSignedHash.Length);
+            using (IncrementalHash hasher = IncrementalHash.CreateHash(hashAlgorithm))
+            {
+                int bufferLength = ComputeToBeSignedEncodedSize(SigStructureCoxtextSign1, _protectedHeaderAsBstr, ReadOnlySpan<byte>.Empty);
+                byte[] buffer = ArrayPool<byte>.Shared.Rent(bufferLength);
 
-            return VerifyHash(key, hashAlgorithm, toBeSignedHash, keyType);
+                try
+                {
+                    await CreateHashedToBeSignedAsync(buffer, hasher, SigStructureCoxtextSign1, _protectedHeaderAsBstr, content, hashAlgorithm).ConfigureAwait(false);
+                    return VerifyHash(key, hasher, hashAlgorithm, keyType);
+                }
+                finally
+                {
+                    ArrayPool<byte>.Shared.Return(buffer, clearArray: true);
+                }
+            }
+        }
+
+        [UnsupportedOSPlatform("browser")]
+        private bool VerifyHash(AsymmetricAlgorithm key, IncrementalHash hasher, HashAlgorithmName hashAlgorithm, KeyType keyType)
+        {
+#if NETSTANDARD2_0 || NETFRAMEWORK
+            byte[] hash = hasher.GetHashAndReset();
+            if (_content != null)
+            {
+                _toBeSignedHash = hash;
+            }
+#else
+            Span<byte> hash = stackalloc byte[0];
+            if (_content != null)
+            {
+                hash = _toBeSignedHash = hasher.GetHashAndReset();
+            }
+            else
+            {
+                hash = stackalloc byte[hasher.HashLengthInBytes];
+                Debug.Assert(hasher.HashLengthInBytes <= 512 / 8); // largest hash we can get (SHA512).
```

Comment:
The assert needs to be before the stackalloc.  Otherwise a surprisingly big value can produce a StackOverflowException that the assert was supposed to warn you about.